edge/pkg/metamanager/metaserver/handlerfactory/handler.go

```diff
@@ -43,25 +45,36 @@ func NewFactory() Factory {
 		scope:             scope.NewRequestScope(),
 		MinRequestTimeout: 1800 * time.Second,
 		handlers:          make(map[string]http.Handler),
+		lock:              &sync.RWMutex{},
 	}
 	return f
 }
 
 func (f *Factory) Get() http.Handler {
+	f.lock.RLock()
 	if h, ok := f.handlers["get"]; ok {
+		f.lock.RUnlock()
 		return h
 	}
+	f.lock.RUnlock()
+	f.lock.Lock()
 	h := handlers.GetResource(f.storage, f.storage, f.scope)
 	f.handlers["get"] = h
+	f.lock.Unlock()
```

Comment:
Too many locks and unlocks can seem confusing.
Whether the read operation can be placed in a separate function.
eg.
```go

func (f *Factory) getHandler(key string) (http.Handler, bool) {
        f.lock.RLock()
        defer f.lock.RUnlock()

        h, ok := f.handlers[key]
        return h, ok
}

func (f *Factory) Get() http.Handler {
        if h, ok := f.getHandler("get"); ok {
                return h
        }

        f.lock.Lock()
        defer f.lock.Lock()
        h := handlers.GetResource(f.storage, f.storage, f.scope)
        f.handlers["get"] = h
        return h
}
```

Reply:
Yes, after extracting it, it seems that there is less unnecessary confusion.